### DISCUSSION TITLE
Compile with AutoFDO (Automatic Feedback Directed Optimizer)

### DISCUSSION THREAD

**Discussion by tiran** [faster-cpython/ideas#124] [Ideas]
AutoFDO is a sampling-based optimization technique supported by clang and gcc. It uses perf to collect sampling data. FDO seems to have some benefits over PGO. For example it does not require special PGO profiling builds. FDO profiles can be shared and reused even after moderate code changes.

Resources:
* https://gcc.gnu.org/wiki/AutoFDO
* https://gcc.gnu.org/wiki/AutoFDO/Tutorial
* https://clang.llvm.org/docs/UsersManual.html#profile-guided-optimization
* https://linuxplumbersconf.org/event/7/contributions/798/attachments/661/1214/LTO_PGO_and_AutoFDO_-_Plumbers_2020_-_Tolvanen_Wendling_Desaulniers.pdf
* https://elinux.org/images/4/4d/Moll.pdf

**Comment by tiran** [comment]
How to build a recent version of autofdo:

```
sudo dnf install -y gcc gcc-c++ ninja-build openssl-devel protobuf-devel libunwind-devel
git clone --recursive --depth=1 --shallow-submodules https://github.com/google/autofdo.git
# remove 1 GB of testdata
find -name testdata | xargs rm -rf
```

```
mkdir autofdo/build
pushd autofdo/build
cmake -G Ninja DCMAKE_INSTALL_PREFIX=. ../
ninja
popd
```

```
git clone https://github.com/andikleen/pmu-tools
```

```
.../pmu-tools/ocperf.py record -b -e br_inst_retired.near_taken:pp -- ./python -m test --pgo
.../autofdo/create_gcov --binary=./python --profile=perf.data --gcov=python.gcov
```

perf records a good amoung of events (``perf record: Captured and wrote 33,724 MB perf.data (43646 samples)``), but the resulting gcov file is empty (1600 bytes, mostly NULL).

**Comment by markshannon** [comment]
Does FDO ever give better results than full PGO?

**Comment by tiran** [comment]
> Does FDO ever give better results than full PGO?

Yes, examples showed similar to slightly better results than full PGO. For FDO you run unmodified code without instrumentation under ``perf`` and use CPU hardware features to measure behavior. Measurement does not cause a major performance impact and lets you monitor behavior in production. There is no instrumentation that interferes with CPU cache lines. The GCC tutorial links goes into more details.

Reuse of FDO profiles sounds appealing to me, too. Downside: I cannot get it work (yet).

**Comment by markshannon** [comment]
Try definitely give this a try.
@tiran do let us now if you ever get it work :slightly_smiling_face:
